### PR TITLE
[temp][fsdp2] infer spmd mesh from global mesh

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126897

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 1]
```diff
@@ -199,21 +200,48 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
...
                 raise NotImplementedError(
                     f"FSDP only supports 1D TP, not {self._tp_spec.placements}"
                 )
             global_placements: List[Placement] = [Replicate(), Replicate()]
             global_dp_mesh_dim = _mesh_resources.get_parent_mesh_dim(dp_mesh)
             global_tp_mesh_dim = _mesh_resources.get_parent_mesh_dim(tp_mesh)
+
+            def infer_spmd_mesh(global_mesh, global_dp_mesh_dim, global_tp_mesh_dim):
+                from torch.distributed.device_mesh import _mesh_resources, DeviceMesh
+                pp_submesh_tensors = global_mesh.mesh.unbind(0)
```
Comment (wconstab) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:213]: what does this do?

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 2]
```diff
@@ -199,21 +200,48 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
...
+                spmd_mesh_dim_names = (
+                    global_mesh.mesh_dim_names[global_dp_mesh_dim],
+                    global_mesh.mesh_dim_names[global_tp_mesh_dim],
+                )
+                for spmd_mesh_tensor in pp_submesh_tensors:
```
Comment (wconstab) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:218]: i couldn't decide if `spmd_mesh_tensor` represents 'some tensor we created on the old mesh' or 'a tensor that represents the ranks that are in a submesh'

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:213]: it's basically generate from a 3-D mesh to a list of 2D meshes. i.e. 
```
>>> import torch
>>> a = torch.arange(8).reshape(2, 2, 2)
>>> a
tensor([[[0, 1],
         [2, 3]],

        [[4, 5],
         [6, 7]]])
>>> a.unbind(0)
(tensor([[0, 1],
        [2, 3]]), tensor([[4, 5],
        [6, 7]]))
```

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 2 above)
Comment (wconstab) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:218]: ok i think i see whats going on.  maybe global mesh starts like 
```
tensor([[[0, 1],
         [2, 3]],

        [[4, 5],
         [6, 7]]])
```

and unbind(0) deletes the outer dim and returns two submesh tensors that looks like 
```
tensor([[0, 1],
        [2, 3]])

tensor([[4, 5],
        [6, 7]])
```

then you check which one your global rank is in and thats your tensor

Comment (wanchaol) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:218]: yes!

Comment (wconstab) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:218]: but this seems like a hard way to accomplish the goal.  what happens if you just slice out a submesh instead?

could we support something like 
```
world_mesh[("dp", "tp")]
[rank0]:(Pdb) [rank0]:*** KeyError: ("Mesh dimension '('dp', 'tp')' does not exist.", "Available mesh dimensions are: mesh_dim_names=('pp', 'dp', 'tp')")
```

Comment (wanchaol) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:218]: Yeah I think that would be better to live in FSDP code, what I implemented here is essentially the submesh slicing implementation logic. cc @wz337

Comment (wconstab) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:218]: > Yeah I think that would be better to live in FSDP code,

you mean, better to live in device-mesh code?

Comment (wconstab) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:218]: oh, ok. got it. yea, agreed

**Pull request closed without merging**